src/parsing.jl

```diff
@@ -1,156 +1,241 @@
-# Functions to parse strings to intervals
-
 """
     parse(DecoratedInterval, s::AbstractString)
 
-Parse a string of the form `"[a, b]_dec"` as a `DecoratedInterval`
-with decoration `dec`.
+Parse a string of the form `"[a, b]_dec"` as a `DecoratedInterval` with decoration `dec`.
+If the decoration is not specified, it is computed based on the parsed interval.
+If the input is an invalid string, a warning is printed and [NaI] is returned. The parser is
+case unsensitive.
+
+### Examples
+
+```jldoctest
+julia> @format true
+Display parameters:
+- format: standard
+- decorations: true
+- significant figures: 6
+
+julia> parse(DecoratedInterval{Float64}, "[1, 2]")
+[1, 2]_com
+
+julia> parse(DecoratedInterval{Float64}, "[1, 2]_def")
+[1, 2]_def
+
+julia> parse(DecoratedInterval{Float64}, "foobar")
+┌ Warning: invalid input, returning [NaI]
+└ @ IntervalArithmetic ~/.julia/dev/IntervalArithmetic/src/parsing.jl:44
+[NaN, NaN]_ill
+```
 """
 function parse(::Type{DecoratedInterval{T}}, s::AbstractString) where T
-    if '_' ∉ s
-        x = parse(Interval{T}, s)
-        return DecoratedInterval(x.lo, x.hi)
-    end
+    s = lowercase(strip(s))
+    s == "[nai]" && return nai(T)
+    try
+        if '_' ∉ s
+            ival, _ = _parse(Interval{T}, s)
+            return DecoratedInterval(ival)
+        end
+
+        decorations = Dict(
+            "ill" => ill,
+            "trv" => trv,
+            "def" => def,
+            "dac" => dac,
+            "com" => com)
 
-    decorations = Dict(
-        "ill" => ill,
-        "trv" => trv,
-        "def" => def,
-        "dac" => dac,
-        "com" => com)
+        interval_string, dec = split(s, "_")
 
-    interval_string, dec = split(s, "_")
-    interval = parse(Interval{T}, interval_string)
-    return DecoratedInterval(interval, decorations[lowercase(dec)])
+        ival, isnotcom = _parse(Interval{T}, interval_string)
+        dec_calc = decoration(ival)
+
+        haskey(decorations, dec) || throw(ArgumentError("invalid decoration $dec"))
+        dec_given = decorations[dec]
+
+        #=
+            If I try to give a decoration that is too high, e.g. [1, Inf]_com, then it
+            should error and return [NaI]. Exception to this is if the interval would be com
+            but becomes dac because of finite precision, e.g. "[1e403]_com" when parse to
+            Interval{Float64} is allowed to become [prevfloat(Inf), Inf]_dac without erroring.
+            The isnotcom flag returned by _parse is used to track if the interval was originally
+            smaller than com or became dac because of overflow.
+        =#
+        dec_given > dec_calc && isnotcom && throw(ArgumentError("invalid decoration $dec for $ival"))
+
+        return DecoratedInterval(ival, min(dec_given, dec_calc))
```

Comment:
I guess it would be better here to explicitly return `DecoratedInterval{T}(...)`, perhaps (but not sure) for type-stability issues.

Reply:
I don't know :D 
if `ival` is already of type `Interval{T}` then it should be type stable I guess?

Reply:
You are right...